### PR TITLE
LibGfx: Road to OpenType hinting (part 1)

### PR DESCRIPTION
**Pull request by MacDue**
This PR is _maybe_ a bit of a long shot as I don't know how much others share my intrigue here...
But this is my first little baby steps into trying to add an OpenType interpreter. 

This PR defines all the OpenType opcodes and adds some helpers for processing instruction data, and also adds some accessors to `OpenType::Font` to get at the raw instruction data.

With this I've created a `ttfdisasm` utility that allows us to have a peek inside some fonts. It allows for disassembling:

 - The 'fpgm' table, this a program that's run once when the font is loaded. It's used to define instructions and functions used by used by other programs.
 - The 'prep' table, this is a general program that's run when ever the font size (or other properties) changes.
 - And the programs associated with any individual glyph.

(The disassembly format is based off the examples in the specification)

With this you can see that all of the LiberationSerif-XXX.tff fonts in Serenity have these programs ready to go:

Here's some examples of what this looks like:
> image
> A random function definition from `LiberationSerif-Bold.ttf` 
---
> image
> The program associated with the glyph for the character '1' in `LiberationSerif-Italic.tt`.
> You can see this is just calling a function defined elsewhere then manipulating some points with the various shift point instructions (e.g. `SHP[]`).

These programs are stack based (hence why you see a lot of pushing before doing stuff).

### CODE REVIEW

File: Userland/Libraries/LibGfx/Font/OpenType/Font.cpp [hunk 1]
```diff
@@ -489,6 +497,14 @@ ErrorOr<NonnullRefPtr<Font>> Font::try_load_from_offset(ReadonlyBytes buffer, u3
     if (opt_kern_slice.has_value())
         kern = TRY(Kern::from_slice(opt_kern_slice.value()));
 
+    Optional<Fpgm> fpgm;
+    if (opt_fpgm_slice.has_value())
+        fpgm = Fpgm(opt_fpgm_slice.value());
```
Comment (ldm5180) [inline Userland/Libraries/LibGfx/Font/OpenType/Font.cpp:502]: Question: why are these not extracted into functions similar to `Font::font_program()` and `Font::control_value_program()`?

Comment (MacDue) [reply Userland/Libraries/LibGfx/Font/OpenType/Font.cpp:502]: I'm not sure what you mean? This is the initial loading of the font where the tables are first extracted, it's only ever run once when the font is loaded. It is a little oddly factored but this is pretty much what the code looks like for all tables, and I didn't want to start refactoring things in the PR. (See the how the OS/2 and Kern tables are loaded just above).

File: Userland/Libraries/LibGfx/Font/OpenType/Tables.h [hunk 2]
```diff
@@ -171,6 +171,36 @@ struct GlyphHorizontalMetrics {
     i16 left_side_bearing;
 };
 
+// https://learn.microsoft.com/en-us/typography/opentype/spec/fpgm
+// fpgm: Font Program
+struct Fpgm {
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/Font/OpenType/Tables.h:176]: Reading fpgm/prep in a commit about adding accessors? rebase mixup perhaps?

**Review by alimpfard — COMMENTED** [review]
just a couple tiny comments
also, ttf debugger :eyes:

File: Userland/Libraries/LibGfx/Font/OpenType/Hinting/Opcodes.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    return (to_underlying(m_opcode) >> (m_flag_bits - 5)) & 1;
+}
+
+void InstructionStream::process_next_instruction()
+{
+    auto opcode = static_cast<Opcode>(next_byte());
+    auto& stream = *this;
+    m_handler.before_operation(stream, opcode);
+    ScopeGuard after = [=, this]() mutable {
+        m_handler.after_operation(stream, opcode);
```
[the remark this answers is not in the document]
Comment (MacDue) [reply Userland/Libraries/LibGfx/Font/OpenType/Hinting/Opcodes.cpp:75]: Nope :sweat_smile:

File: Userland/Libraries/LibGfx/Font/OpenType/Tables.h (hunk 2 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/Font/OpenType/Tables.h:176]: Ah, I see where @ldm5180 was coming from now

Comment (MacDue) [reply Userland/Libraries/LibGfx/Font/OpenType/Tables.h:176]: (Split this out into two commits now)

**Pull request merged**